lls/nd.c

```diff
@@ -0,0 +1,739 @@
+	ND_OPT_SOURCE_LL_ADDR = 1,
+	/* Link-layer address of the target, optional in Advertisements. */
+	ND_OPT_TARGET_LL_ADDR = 2,
+	/* Other options exist but are not supported here. */
+	ND_OPT_MAX,
+};
+
+struct nd_opts {
+	/* Pointers to each option present in an ICMPv6 packet. */
+	struct nd_opt_hdr *opt_array[ND_OPT_MAX];
+};
+
+struct nd_opt_hdr {
+	/* Type of the option. */
+	uint8_t type;
+	/* Length of option (including @type and @len) in units of 64 bits. */
+	uint8_t len;
+} __attribute__((__packed__));
+
+/* Used for both ND_OPT_SOURCE_LL_ADDR and ND_OPT_TARGET_LL_ADDR. */
+struct nd_opt_lladdr {
+	/* Type of the option. */
+	uint8_t           type;
+	/* Length of option (including @type and @len) in units of 64 bits. */
+	uint8_t           len;
+	/* Hardware address corresponding to @type. */
+	struct ether_addr ha;
+} __attribute__((__packed__));
+
+/* Minimum size of a Neighbor Discovery packet with a link-layer option. */
+#define ND_NEIGH_PKT_LLADDR_MIN_LEN (ND_NEIGH_PKT_MIN_LEN + \
+	sizeof(struct nd_opt_lladdr))
+
+/* Flags for Neighbor Advertisements. */
+#define LLS_ND_NA_ROUTER    0x80000000
+#define LLS_ND_NA_SOLICITED 0x40000000
+#define LLS_ND_NA_OVERRIDE  0x20000000
+
+/*
+ * Returns whether a given IPv6 address is the unspecified address,
+ * which can be used for duplicate address detection.
+ */
+static inline int
+ipv6_addr_unspecified(const uint8_t *ip6_addr)
+{
+	const uint64_t *paddr = (const uint64_t *)ip6_addr;
+	return (paddr[0] | paddr[1]) == 0ULL;
+}
+
+/*
+ * Returns whether a given IPv6 address is a solicited-node
+ * multicast address.
+ */
+static inline int
+ipv6_addr_solicited_node_mc(const uint8_t *ip6_addr)
+{
+	const uint64_t *paddr = (const uint64_t *)ip6_addr;
+	return ((paddr[0] ^ rte_cpu_to_be_64(0xff02000000000000UL)) |
+		((paddr[1] ^ rte_cpu_to_be_64(0x00000001ff000000UL)) &
+		rte_cpu_to_be_64(0xffffffffff000000UL))) == 0ULL;
+}
+
+/* Returns whether a given IPv6 address is generally a multicast address. */
+static inline int
+ipv6_addr_multicast(const uint8_t *ip6_addr)
+{
+	return ((*(const uint32_t *)ip6_addr) & rte_cpu_to_be_32(0xFF000000)) ==
+		rte_cpu_to_be_32(0xFF000000);
+}
+
+int
+iface_nd_enabled(struct net_config *net, struct gatekeeper_if *iface)
+{
+	/* When @iface is the back, need to make sure it's enabled. */
+	if (iface == &net->back)
+		return net->back_iface_enabled &&
+			iface->configured_proto & GK_CONFIGURED_IPV6;
+
+	/* @iface is the front interface. */
+	return iface->configured_proto & GK_CONFIGURED_IPV6;
+}
+
+char *
+ipv6_str(struct lls_cache *cache, const uint8_t *ip_be, char *buf, size_t len)
+{
+	struct in6_addr ipv6_addr;
+
+	if (sizeof(ipv6_addr) != cache->key_len) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: the key size of an ND entry should be %zu, but it is %"PRIx32"\n",
+			sizeof(ipv6_addr), cache->key_len);
+		return NULL;
+	}
+
+	rte_memcpy(ipv6_addr.s6_addr, ip_be, sizeof(ipv6_addr.s6_addr));
+	if (inet_ntop(AF_INET6, &ipv6_addr, buf, len) == NULL) {
```

Comment:
Why didn't you just pass `ip_be` to `inet_ntop()` since its second argument is `const void *`?